lib/app/modules/smart/smart_discovery_sequence.rb

```diff
@@ -10,129 +10,216 @@ class SMARTDiscoverySequence < SequenceBase
+        'revocation_endpoint'
+      ].freeze
+
+      SMART_OAUTH_EXTENSION_URL = 'http://fhir-registry.smarthealthit.org/StructureDefinition/oauth-uris'
+
+      REQUIRED_OAUTH_ENDPOINTS = [
+        { url: 'authorize', description: 'authorization' },
+        { url: 'token', description: 'token' }
+      ].freeze
+
+      OPTIONAL_OAUTH_ENDPOINTS = [
+        { url: 'register', description: 'dynamic registration' },
+        { url: 'manage', description: 'authorization management' },
+        { url: 'introspect', description: 'token introspection' },
+        { url: 'revoke', description: 'token revocation' }
+      ].freeze
+
+      test 'Retrieve Configuration from well-known endpoint' do
         metadata do
           id '01'
           link 'http://www.hl7.org/fhir/smart-app-launch/conformance/#using-well-known'
           desc %(
-            The authorization endpoints accepted by a FHIR resource server can be exposed as a Well-Known Uniform Resource Identifier
+            The authorization endpoints accepted by a FHIR resource server can
+            be exposed as a Well-Known Uniform Resource Identifier
           )
         end
 
-        oauth_configuration_url = @instance.url.chomp('/') + '/.well-known/smart-configuration'
-        oauth_configuration_response = LoggedRestClient.get(oauth_configuration_url)
-        assert_response_ok(oauth_configuration_response)
-        oauth_configuration_response_body = JSON.parse(oauth_configuration_response.body)
-
-        assert !oauth_configuration_response_body.nil?, 'No authorization response body available'
-        @well_known_authorize_url = oauth_configuration_response_body['authorization_endpoint']
-        @well_known_token_url = oauth_configuration_response_body['token_endpoint']
-        capabilities = oauth_configuration_response_body['capabilities']
-        register_url = oauth_configuration_response_body['registration_endpoint']
+        well_known_configuration_url = @instance.url.chomp('/') + '/.well-known/smart-configuration'
+        well_known_configuration_response = LoggedRestClient.get(well_known_configuration_url)
+        assert_response_ok(well_known_configuration_response)
+        assert_response_content_type(well_known_configuration_response, 'application/json')
+        assert_valid_json(well_known_configuration_response.body)
+
+        @well_known_configuration = JSON.parse(well_known_configuration_response.body)
+        @well_known_authorize_url = @well_known_configuration['authorization_endpoint']
+        @well_known_token_url = @well_known_configuration['token_endpoint']
+        @well_known_register_url = @well_known_configuration['registration_endpoint']
+        @well_known_manage_url = @well_known_configuration['management_endpoint']
+        @well_known_introspect_url = @well_known_configuration['introspection_endpoint']
+        @well_known_revoke_url = @well_known_configuration['revocation_endpoint']
+
+        @instance.update(
```

Comment:
It seems that we should be saving the SMART .well-known file like we save the capability statement.  I was expecting to see some of this info show up in the state modal, but we are only saving a few of these fields.  Push this off to investigate in another ticket?

Reply:
Seems like a good idea to me.